lib/gsup/controller/ulr.py

```
@@ -132,27 +141,40 @@ async def handle_message(self, peer: IPAPeer, message: GsupMessage):
             imsi = GsupMessageUtil.get_first_ie_by_name('imsi', request_dict)
             if imsi is None:
                 raise ValueError(f"Missing IMSI in GSUP message from peer {peer}")
-            validate_imsi(imsi)
-            subscriber_info = self._database.Get_Gsup_SubscriberInfo(imsi)
+            try:
+                validate_imsi(imsi)
+            except InvalidIMSI as e:
+                raise ULRError(f"Invalid IMSI: {imsi}", GMMCause.INV_MAND_INFO) from e
+
+            rat_type = GsupMessageUtil.get_first_ie_by_name('current_rat_type', request_dict)
+            rat_type_to_check = RAT.GERAN  # Default to 2G if not specified
+
+            # Current RAT Type is a list for some reason. Maybe a bug in osmocom?
+            if rat_type is not None and rat_type[0] == 'utran':
+                rat_type_to_check = RAT.UTRAN
+
+            try:
+                subscriber_info = self._database.Get_Gsup_SubscriberInfo(imsi)
+                subscriber = self._database.Get_Subscriber(imsi=imsi, get_attributes=True)
+            except ValueError as e:
+                raise ULRError(f"Subscriber not found: {imsi}", GMMCause.IMSI_UNKNOWN) from e
```

Note: IMHO it would make sense to define a SubscriberNotFound exception and use it here instead of using the generic ValueError exception as otherwise this can mask other errors and return a misleading error message. I ran into this last week with a similar code path. This doesn't need to block the PR though since this was just moved around.